kernel: add support for Intel(R) 10/40GbE PCI Express x710 adapters and Virtual Function devices

peterwillcn opened the pull request:
Add Support for Intel(R) 10/40GbE PCI Express x710 adapters: 
- kmod-i40e	4.14.101-1	Kernel modules for Intel(R) Ethernet Controller XL710 Family 10/40 Gigabit Ethernet adapters.	
- kmod-i40evf	4.14.101-1	Kernel modules for Intel(R) Ethernet Controller XL710 Family Virtual Function Ethernet adapters.

```
[  3.359563] i40e: Intel(R) Ethernet Connection XL710 Network Driver - version 2.1.14-k
[  3.374535] i40e: Copyright (c) 2013 - 2014 Intel Corporation.
[  3.376581] i40evf: Intel(R) 10 Gigabit PCI Express Virtual Function Network Driver - version 2.1.14-k
[  3.377524] i40evf: Copyright (c) 2009 - 2015 Intel Corporation.
[  3.378532] i40evf 0000:00:05.0: 00:11:22:33:93:01
[  3.379527] i40evf 0000:00:05.0: MAC: 4
[  3.380096] i40evf 0000:00:05.0: Intel(R) X710 Virtual Function
[  3.384642] i40evf 0000:00:06.0: 00:11:22:33:93:02
[  3.389573] i40evf 0000:00:06.0: MAC: 4
[  3.392017] i40evf 0000:00:06.0: Intel(R) X710 Virtual Function
```

Compile-tested on: x86_64
Runtime-tested on: x86_64

Signed-off-by: Xiaobo Tian <peterwillcn@gmail.com>

hauke commented:
Please use your real name in the Signed-off-by: line

peterwillcn commented:
@hauke xiaobo is my real name in China.

peterwillcn commented:
@jow Ping

peterwillcn commented:
@jow- ping

chunkeey commented:
The subject "kernel: add support for Intel(R) Ethernet XL710, X710, X722 Adaptive Virtual Function devices" will look very weird. Do you think you can keep [it somewhere around the 50 character limit](https://openwrt.org/submitting-patches)?

chunkeey commented:
> @hauke xiaobo is my real name in China.

https://github.com/panyanyany/learn_eos/blob/master/dawn-2.x/Docker/full.Dockerfile

lists "Xiaobo Tian <peterwillcn@gmail.com>" ?

peterwillcn committed:
kernel: support intel X7xx 10/40GbE adapters

Add Support for Intel(R) 10/40GbE PCI Express x710 adapters:

* kmod-i40e	4.14.101-1	Kernel modules for Intel(R) Ethernet Controller XL710 Family 10/40 Gigabit Ethernet adapters.
* kmod-i40evf	4.14.101-1	Kernel modules for Intel(R) Ethernet Controller XL710 Family Virtual Function Ethernet adapters.

```
[  3.359563] i40e: Intel(R) Ethernet Connection XL710 Network Driver - version 2.1.14-k
[  3.374535] i40e: Copyright (c) 2013 - 2014 Intel Corporation.
[  3.376581] i40evf: Intel(R) 10 Gigabit PCI Express Virtual Function Network Driver - version 2.1.14-k
[  3.377524] i40evf: Copyright (c) 2009 - 2015 Intel Corporation.
[  3.378532] i40evf 0000:00:05.0: 00:11:22:33:93:01
[  3.379527] i40evf 0000:00:05.0: MAC: 4
[  3.380096] i40evf 0000:00:05.0: Intel(R) X710 Virtual Function
[  3.384642] i40evf 0000:00:06.0: 00:11:22:33:93:02
[  3.389573] i40evf 0000:00:06.0: MAC: 4
[  3.392017] i40evf 0000:00:06.0: Intel(R) X710 Virtual Function
```
Compile-tested on: ar71xx, cns3xxx, imx6, x86_64
Runtime-tested on: ar71xx, cns3xxx, imx6, x86_64

Signed-off-by: Xiaobo Tian <peterwillcn@gmail.com>

peterwillcn commented:
@chunkeey  correct

chunkeey commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -642,6 +642,39 @@ endef
+  KCONFIG:=CONFIG_I40E \
+    CONFIG_I40E_DCB=n
+  FILES:=$(LINUX_DIR)/drivers/net/ethernet/intel/i40e/i40e.ko
+  AUTOLOAD:=$(call AutoLoad,36,i40e)
```
Is there a reason for AutoLoad? Because AutoProbe is definitely preferred because it only loads the driver if there's really a device. 

(The same goes for the i40evf)

peterwillcn replied:
I40e is Physical Functions drivers
i40evf is Virtual Functions drivers
Replace AutoLoad with AutoProbe @ 474061cf82

chunkeey replied:
You don't need to replace the autoload for devices you don't have and can't test. Please focus on your i40e+vf instead.  It would be great if you could squash the [changes into one commit](https://openwrt.org/submitting-patches#squashing_commits).

Also, please add your github PR's text into the patch commit message as well.
(reason being: only the patches will be merged, so the extensive tested on
list will be lost)

peterwillcn replied:
OK

chunkeey commented:
While compiling, I ran into this issue:

```
drivers/net/ethernet/intel/i40e/i40e_ethtool.c: In function 'i40e_set_priv_flags':
drivers/net/ethernet/intel/i40e/i40e_ethtool.c:4160:6: error: implicit declaration of function 'cmpxchg64'; did you mean 'cmpxchg'? [-Werror=implicit-function-declaration]
  if (cmpxchg64(&pf->flags, orig_flags, new_flags) != orig_flags) {
      ^~~~~~~~~
      cmpxchg
cc1: some warnings being treated as errors
```

This happen on apm821xx and mpc85xx (PowerPC) due to cmpxchg64 being gated there unless it's a PPC64. From what I can tell, this seems to be fixed with 4.15+ , do you think you can backport the changes?

peterwillcn committed:
kernel upgrade to 4.19
```
[    0.000000] Linux version 4.19.24 (openwrt@xxx) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9420-c17a68c)) #0 SMP Sat Feb 23 01:58:20 2019
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz root=PARTUUID=82ee4abc-02 rootfstype=ext4 rootwait console=tty0 console=ttyS0,115200n8 noinitrd
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
```
Compile-tested on: x86_64
Runtime-tested on: x86_64

Signed-off-by: Xiaobo Tian <peterwillcn@gmail.com>

peterwillcn commented:
Opened a new pull request https://github.com/openwrt/openwrt/pull/1851
